cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+                    bit_width: bw_lhs,
+                    max: max_lhs,
+                },
+                Fact::ValueMax {
+                    bit_width: bw_rhs,
+                    max: max_rhs,
+                },
+            ) => {
+                // If the bitwidths we're claiming facts about are the
+                // same, and if the value is less than or equal to
+                // `max_lhs`, and if `max_rhs` is less than `max_lhs`,
+                // then it is certainly less than or equal to
+                // `max_rhs`.
+                //
+                // In other words, we can always expand the claimed
+                // possible value range.
+                bw_lhs == bw_rhs && max_lhs <= max_rhs
+            }
+            (
+                Fact::PointsTo {
+                    region: MemoryRegion { max: max_lhs },
+                },
+                Fact::PointsTo {
+                    region: MemoryRegion { max: max_rhs },
+                },
+            ) => {
+                // If the pointer is valid up to `max_lhs`, and
+                // `max_rhs` is less than or equal to `max_lhs`, then
+                // it is certainly valid up to `max_rhs`.
+                //
+                // In other words, we can always shrink the valid
+                // addressable region.
+                max_rhs <= max_lhs
+            }
+            _ => false,
+        }
+    }
+
+    /// Computes whatever fact can be known about the sum of two other
+    /// values with attached facts. The add is performed to the given
+    /// bit-width, and the target machine has the given pointer-width.
+    pub fn add(&self, other: &Self, add_width: u16, pointer_width: u16) -> Option<Self> {
+        match (self, other) {
+            (
+                Fact::ValueMax {
+                    bit_width: bw_lhs,
+                    max: lhs,
+                },
+                Fact::ValueMax {
+                    bit_width: bw_rhs,
+                    max: rhs,
+                },
+            ) if bw_lhs == bw_rhs && add_width >= *bw_lhs => {
+                let computed_max = lhs.checked_add(*rhs)?;
+                Some(Fact::ValueMax {
+                    bit_width: *bw_lhs,
+                    max: computed_max,
+                })
+            }
+
+            (
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+                Fact::PointsTo { region },
+            )
+            | (
+                Fact::PointsTo { region },
+                Fact::ValueMax {
+                    bit_width: bw_max,
+                    max,
+                },
+            ) if *bw_max >= pointer_width && add_width >= *bw_max => {
+                let computed_region = MemoryRegion {
+                    max: region.max.checked_sub(*max)?,
+                };
+                Some(Fact::PointsTo {
+                    region: computed_region,
+                })
```

Comment:
Stylistic nitpick: I'd personally either name `computed_region` just `region` so we can use short-hand struct literal syntax for the field, or even inline the variable into the struct literal field.

Reply:
Done!